pkg/lease/client.go

```diff
@@ -0,0 +1,44 @@
+package lease
+
+import (
+	"log"
+
+	boskos "k8s.io/test-infra/boskos/client"
+)
+
+const (
+	freeState   = "free"
+	leasedState = "leased"
+)
+
+type Client interface {
+	Acquire(rtype string) (string, error)
+	Heartbeat() error
+	Release(name string) error
+	ReleaseAll() error
+}
+
+func NewClient(owner, url string) Client {
+	return &client{*boskos.NewClient(owner, url)}
+}
+
+type client struct{ boskos boskos.Client }
+
+func (c *client) Acquire(rtype string) (string, error) {
+	r, err := c.boskos.Acquire(rtype, freeState, leasedState)
+	if err != nil {
+		return "", err
+	}
+	return r.Name, nil
+}
+
+func (c *client) Heartbeat() error          { return c.boskos.UpdateAll(leasedState) }
+func (c *client) Release(name string) error { return c.boskos.ReleaseOne(name, freeState) }
+
+func (c *client) ReleaseAll() error {
+	if !c.boskos.HasResource() {
+		return nil
+	}
+	log.Println("error: Not all leases were released, releasing them now")
```

Comment:
I don't think this is the proper place to log this error - if we can easily do `ReleaseAll`, why complain about it in library code? If the caller sets a rule for itself that each rule needs to be individually released, why not, but I don't see a reason why the lib should do that.

Reply:
Yea, let's return the error here.

Reply:
So are you proposing to also expose `HasResource` to allow the caller to do the check?

Would it be better if the method had a name that made it clearer that we are checking for leaks?

Reply:
> So are you proposing to also expose `HasResource` to allow the caller to do the check?

Depends on that is the API's contract: if it imposes "all leases need to eventually be explicitly released", then no, but that contract also seems quite strong for library that has easy `ReleaseAll` and so can't really leak.

> Would it be better if the method had a name that made it clearer that we are checking for leaks?

I think this would be best: make it more clear that releases should be explicit and `ReleaseAll` is really the last resort call that is usually not expected to really find anything not released. Maybe drop the severity of the message to "warn".